test/conftest.py

```diff
@@ -40,16 +41,68 @@ def fixture_trace(request):
     trace.tracer_state = _TracerFSM.INITIAL
 
 
+@pytest.fixture(scope="function")
+def check_tracer_events():
+    def serialize_symbol(symbol):
+        if symbol.snapshot is None:
+            return {"name": symbol.name, "snapshot": symbol.snapshot}
+        snapshot = {
+            "location": symbol.snapshot.location,
+            "events_pointer": symbol.snapshot.events_pointer,
+        }
+        return {"name": symbol.name, "snapshot": snapshot}
+
+    yield
+
+    tracer_events = []
+    for event in trace.events:
+        event_dict = attr.asdict(event)
+        for key, val in event_dict.items():
+            if type(val) == Symbol:
+                event_dict[key] = serialize_symbol(val)
+            elif type(val) == list:
+                event_dict[key] = sorted(
+                    [serialize_symbol(sym) for sym in val if type(sym) == Symbol],
+                    key=lambda x: x["name"],
+                )
+        event_dict["__class__"] = event.__class__.__name__
+        tracer_events.append(event_dict)
+
+    # Generates golden data.
+    golden_filepath = f"test/data/{python_version}/{trace.frame.frame_name}.json"
+    directory = os.path.dirname(golden_filepath)
+    if not os.path.exists(directory):
+        os.makedirs(directory)
+
+    if not os.path.exists(golden_filepath):
+        with open(golden_filepath, "wt") as f:
+            json.dump({"tracer.events": tracer_events}, f, ensure_ascii=False, indent=4)
+        return
+
+    # Assuming run in root directory.
+    with open(golden_filepath, "rt") as f:
+        golden_frame_data = json.loads(f.read())
+
+    if golden_frame_data.get("tracer.events", None) is None:
+        with open(golden_filepath, "wt") as f:
+            golden_frame_data["tracer.events"] = tracer_events
+            json.dump(golden_frame_data, f, ensure_ascii=False, indent=4)
+
+    assert tracer_events == golden_frame_data["tracer.events"], json.dumps(
+        tracer_events, indent=4
+    )
+
+
 @pytest.fixture
-def mocked_responses(request):
+def check_response(request):
```

Comment:
I feel like this function is doing multiple things and we should break into smaller ones, something like

```python
golden_response = get_golden_response()
if golden_response:
  assert ...
else:
  update_golden_response()
```

This makese the intent more clear. A good side effect: skip assertion if golden_response does not exist (now assertion still happens if the file exists).

Reply:
Same with `check_tracer_events`